packages/alfa-dom/src/tables/structure.ts

```diff
@@ -0,0 +1,353 @@
+export type RowGroup = {
+  // First row of the group
+  anchor: { y: number };
+  height: number;
+  element: Element;
+}
+
+// https://html.spec.whatwg.org/multipage/tables.html#concept-column-group
+export type ColGroup = {
+  // First column of the group
+  anchor: { x: number };
+  width: number;
+  element: Element;
+}
+
+// export const isCoveredBy: Predicate<Slot, Slot, Array<Cell | RowGroup | ColGroup>> = (slot, cover) => {
+//   if ("width" in cover) { // Cell or Col
+//     if (slot.x < cover.anchor.x) { // slot is left of cover
+//       return false;
+//     }
+//     if (cover.anchor.x + cover.width - 1 < slot.x) { // slot is right of cover
+//       return false;
+//     }
+//   }
+//
+//   if ("height" in cover) { // Cell or Row
+//     if (slot.y < cover.anchor.y) { // slot is above cover
+//       return false;
+//     }
+//     if (cover.anchor.y + cover.height - 1 < slot.y) { // slot is below cover
+//       return false;
+//     }
+//   }
+//
+//   return true;
+// };
+
+// micro syntaxes to move to alfa-parser
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#signed-integers
+export function parseInteger(str: string): Result<readonly [string, number], string> {
+  // empty/whitespace string are errors for specs, not for Number…
+  // \s seems to be close enough to "ASCII whitespace".
+  if (str.match(/^\s*$/)) {
+    return Err.of("The string is empty");
+  }
+  const raw = Number(str);
+  return isNaN(raw) ?
+    Err.of("The string does not represent a number") :
+    raw !== Math.floor(raw) ?
+      Err.of("The string does not represent an integer") :
+      // 0 and -0 are different (but equal…) floats. Normalising.
+      Ok.of(["", raw === -0 ? 0 : raw] as const);
+}
+
+// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-non-negative-integers
+export function parseNonNegativeInteger(str: string): Result<readonly [string, number], string> {
+  const result = parseInteger(str);
+  return result.andThen(([_, value]) =>
+    value < 0 ?
+      Err.of("This is a negative number") :
+      result)
+}
+// end micro syntaxes
+
+// attribute helper should move to attribute
+export function parseAttribute<RESULT, ERROR>(parser: Parser<string, RESULT, ERROR>): Mapper<Attribute, Result<RESULT, ERROR>> {
+  return (attribute) =>
+    parser(attribute.value)
+      .andThen(([_, value]) => Ok.of(value));
+}
+// end attribute helper
+
+export function parseSpan(element: Element, name: string, min: number, max: number, failed: number): number {
+  return element.attribute(name)
+    .map(parseAttribute(parseNonNegativeInteger))
+    .map(r => r.map(x => clamp(x, min, max)))
+    .getOr(Ok.of(failed))
+    .getOr(failed);
+}
+
+// https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-growing-downward-growing-cells
+function growingCell(yCurrent: number, keepGrowing: boolean = false): ((cell: Cell) => void) {
```

Comment:
Needs to update cell (rather than create a new one) because the cell is shared between the table and the slots. Not super fan of it, I’d rather have this as a pure function and repalce the `forEach` with `map` where needed…